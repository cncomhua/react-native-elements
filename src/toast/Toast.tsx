import React, { useContext } from 'react';
import { View, StyleSheet } from 'react-native';
import { SafeAreaInsetsContext } from 'react-native-safe-area-context';

import { ToastContext, ToastPosition } from './ToastProvider';
import Message from './Message';

import type { VFC, Dispatch, SetStateAction } from 'react';
import type { MessageState } from './ToastProvider';

type ToastProps = {
  messages: MessageState[];
  setMessage: Dispatch<SetStateAction<MessageState[]>>;
};

const Toast: VFC<ToastProps> = ({ messages, setMessage }) => {
  if (!messages || !Array.isArray(messages)) {
    throw new Error(
      'Messages is a required prop type of array of MessageState!'
    );
  }

  if (!setMessage || typeof setMessage !== 'function') {
    throw new Error('setMessages is a required prop type of function!');
  }
  const insets = useContext(SafeAreaInsetsContext);
  const { position, containerToastStyle } = useContext(ToastContext);

  const onHideHandler = (id: string): void => {
    setMessage((prevMessages) =>
      prevMessages.filter((currentMessage) => currentMessage.id !== id)
    );
  };

  return (
    <View
      testID={'toast-container-test'}
      style={StyleSheet.flatten([
        {
          top: position === ToastPosition.top ? insets?.top ?? 0 : undefined,
          bottom:
            position === ToastPosition.bottom ? insets?.bottom ?? 0 : undefined,
        },
        containerToastStyle,
      ])}
    >
      {messages.map((message) => (
        <Message
          key={message.id}
          message={message}
          onHide={() => onHideHandler(message.id)}
        />
      ))}
    </View>
  );
};

export default Toast;
