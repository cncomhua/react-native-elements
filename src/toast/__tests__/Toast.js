import React, { useState } from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Toast from '../Toast';
import { act, create } from 'react-test-renderer';
import { Animated } from 'react-native';

describe('Toast container component', () => {
  jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

  let component;

  beforeEach(() => {
    component = null;
  });

  it('should render toast container properly and match snapshot', () => {
    const WrapperComponentWithState = () => {
      const [messages, setMessages] = useState([
        {
          id: 0,
          text: 'Im info text',
          type: 'info',
        },
        {
          id: 1,
          text: 'Im error text',
          type: 'error',
        },
        {
          id: 2,
          text: 'Im warning text',
          type: 'warning',
        },
        {
          id: 3,
          text: 'Im success text',
          type: 'success',
        },
      ]);

      return <Toast messages={messages} setMessage={setMessages} />;
    };

    component = shallow(<WrapperComponentWithState />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render toast container with no messages', () => {
    const WrapperComponentWithState = () => {
      const [messages, setMessages] = useState([]);

      return <Toast messages={messages} setMessage={setMessages} />;
    };

    component = create(<WrapperComponentWithState />);

    expect(component.root.children.length).toBe(1);

    const toastComponent = component.root.children[0];

    expect(toastComponent.findAllByType(Animated.View).length).toBe(0);
  });

  it('should render toast container with handling messages', async () => {
    const WrapperComponentWithState = () => {
      const [messages, setMessages] = useState([
        {
          id: 0,
          text: 'Message 1',
          type: 'info',
        },
        {
          id: 1,
          text: 'Message 2',
          type: 'success',
        },
      ]);

      return <Toast messages={messages} setMessage={setMessages} />;
    };

    act(() => {
      component = create(<WrapperComponentWithState />);
    });

    expect(component.root.findAllByType(Animated.View).length).toBe(2);

    act(() => {
      jest.runAllTimers();
    });

    expect(component.root.findAllByType(Animated.View).length).toBe(0);
  });

  it('should throw error with wrong props', async () => {
    expect(() => shallow(<Toast />)).toThrowError();
    expect(() => shallow(<Toast messages={[1, 2, 3]} />)).toThrowError();
    expect(() => shallow(<Toast setMessage={() => null} />)).toThrowError();
  });
});
