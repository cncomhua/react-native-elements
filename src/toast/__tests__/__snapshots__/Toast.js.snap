// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast container component should render toast container properly and match snapshot 1`] = `
<Toast
  messages={
    Array [
      Object {
        "id": 0,
        "text": "Im info text",
        "type": "info",
      },
      Object {
        "id": 1,
        "text": "Im error text",
        "type": "error",
      },
      Object {
        "id": 2,
        "text": "Im warning text",
        "type": "warning",
      },
      Object {
        "id": 3,
        "text": "Im success text",
        "type": "success",
      },
    ]
  }
  setMessage={[Function]}
/>
`;
